tests/AppTest.php

```diff
@@ -67,4 +68,123 @@ public function testUnexpectedOutputLateError(): void
             ob_end_clean();
         }
     }
+
+    public function provideUrlBuildingCases(): iterable
+    {
+        return [
+            // simple cases
+            [[], false, [], '/', '/index.php', '/default.php8', '/'],
+            [[], false, [], '/test/', '/test/index.php', '/test/default.php8', '/test/'],
+            [[], true, [], '/', '/index.php', '/default.php8', '/'],
+            [[], true, [], '/test/', '/test/index.php', '/test/default.php8', '/test/'],
+
+            // simple cases with query args in request
+            // [[], false, [], '/?test=atk4', '/index.php?test=atk4', '/default.php8?test=atk4', '/?test=atk4'],
```

Comment:
@abbadon1334 why are these tests commented, shouldn't this PR fix all the URL issues? Or is #2101 needed for some reason?

Reply:
No commented out cannot be resolved due to the inconsistency of App::url, In #2101 these tests are uncommented and App::url is refactored to give results consistency.

Reply:
in #2101 we only switch from $_XXX globals to request API, this PR already does some App::url changes, can the changes needed for consistent URLs/all tests passing be done in this PR (using $_XXX globals)?

Reply:
answer: #2101 is really not needed